pilot/pkg/networking/core/v1alpha3/listener.go

```diff
@@ -1633,3 +1639,49 @@ func removeListenerFilterTimeout(listeners []*listener.Listener) {
 func listenerKey(bind string, port int) string {
 	return bind + ":" + strconv.Itoa(port)
 }
+
+const baggageFormat = "k8s.cluster.name=%s,k8s.namespace.name=%s,k8s.%s.name=%s,service.name=%s,service.version=%s"
+
+// outboundTunnelListener is built for each ServiceAccount from pods on the node.
+// This listener adds the original destination headers from the dynamic EDS metadata pass through.
+// We build the listener per-service account so that it can point to the corresponding cluster that presents the correct cert.
+func outboundTunnelListener(push *model.PushContext, proxy *model.Proxy) *listener.Listener {
+	name := "tunnel"
+	canonicalName := proxy.Labels[model.IstioCanonicalServiceLabelName]
+	canonicalRevision := proxy.Labels[model.IstioCanonicalServiceRevisionLabelName]
+	p := &tcp.TcpProxy{
+		StatPrefix:       name,
+		ClusterSpecifier: &tcp.TcpProxy_Cluster{Cluster: name},
+		TunnelingConfig: &tcp.TcpProxy_TunnelingConfig{
+			Hostname: "%DYNAMIC_METADATA(tunnel:destination)%",
+			HeadersToAdd: []*core.HeaderValueOption{
+				{Header: &core.HeaderValue{Key: "x-envoy-original-dst-host", Value: `%DYNAMIC_METADATA(["tunnel", "destination"])%`}},
+
+				{Header: &core.HeaderValue{
+					Key: "baggage",
+					Value: fmt.Sprintf(baggageFormat,
+						proxy.Metadata.ClusterID, proxy.ConfigNamespace,
+						// TODO do not hardcode deployment. But I think we ignore it anyways?
+						"deployment", proxy.Metadata.WorkloadName,
+						canonicalName, canonicalRevision,
+					),
+				}},
+			},
+		},
+	}
+
+	l := &listener.Listener{
+		Name:              name,
+		UseOriginalDst:    wrappers.Bool(false),
+		ListenerSpecifier: &listener.Listener_InternalListener{InternalListener: &listener.Listener_InternalListenerConfig{}},
+		ListenerFilters:   []*listener.ListenerFilter{xdsfilters.SetDstAddress},
+		FilterChains: []*listener.FilterChain{{
+			Filters: []*listener.Filter{setAccessLogAndBuildTCPFilter(push, proxy, p, istionetworking.ListenerClassSidecarOutbound)},
+		}},
+	}
+	accessLogBuilder.setListenerAccessLog(push, proxy, l, istionetworking.ListenerClassSidecarOutbound)
```

Comment:
I have seen multi access logs printed for a single request, not sure is this good for reading

Reply:
Fixed. We do have a TODO in the tracker for improving the logging as well